### ISSUE TITLE
apitrace doesn't create trace files for LWJGL applications on OS X 10.11

### ISSUE THREAD

**Issue by lwjglapitrace** [apitrace/apitrace#491]
I first cloned and built apitrace with:

```
cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_PREFIX_PATH=~/Qt/clang_64
make -C build
```

I then attempted a trace with:

`apitrace trace --api=gl java -Djava.library.path=libs/lwjgl-2.9.1/native/macosx -jar build/libs/app.jar`

My application launched as expected, but apitrace did not create a .trace file

I then attempted another build with:

```
git clone --depth=50 https://github.com/apitrace/apitrace.git apitrace/apitrace
cd apitrace/apitrace
git fetch origin +refs/pull/490/merge:
git checkout -qf FETCH_HEAD

export MAKEFLAGS=-j2
export LABEL="mac64"
export CMAKE_OPTIONS="-DENABLE_GUI=1"

brew update; brew install qt5; export CMAKE_PREFIX_PATH=$(brew --prefix qt5);

cmake -H. -Bbuild $CMAKE_OPTIONS
cmake --build build --use-stderr -- -k
cmake --build build --use-stderr --target check -- -k
cmake --build build --use-stderr --target package -- -k
```

However, using the apitrace binary from the .dmg produced also failed to create a .trace file

I then attempted to create a trace file using the qapitrace GUI, but the GUI failed to produce the trace file:

```
App :  "/usr/bin/java"
  Arguments:  ("-Djava.library.path=/Users/lwjglapitrace/development/app/libs/lwjgl-2.9.1/native/macosx/ -jar /Users/lwjglapitrace/development/app/build/libs/app.jar")
Done tracing  "/Users/lwjglapitrace/development/app/java.trace"
Loading: "/Users/lwjglapitrace/development/app/java.trace"
error: failed to open /Users/lwjglapitrace/development/app/java.trace
```

It's worth noting here that my application was never launched, or rather, no window was ever drawn on screen, but the GUI logs "Done tracing...", as if to indicate that I had closed my app after it had launched, but this is not the case. 

So, I'm not sure if apitrace is failing to parse the arguments supplied, or if this is some kind of permissions issue?

I've also tried explicitly defining the output path of the trace file, but this does not work either

I can use apitrace on Linux to create a .trace file of the same application, so I'm pretty sure this isn't an issue specific to my app

I've also tried specifying the DYLD_LIBRARY_PATH, to point to the wrappers folder (which apitrace successfully picked up on), but this also did not create a .trace file

**Comment by jrfonseca** [comment]
This might be related to issue #409.

I think the problem is that `DYLD_LIBRARY_PATH` mechanism is not very reliable on Mac OS X, particularly with Java apps.

I've started prototyping using `DYLD_INTERPOSE` in https://github.com/apitrace/apitrace/tree/dyld-interpose branch.  But it's still work in progress. Give it a try and let me know if it helps.

**Comment by lwjglapitrace** [comment]
So do I just set DYLD_INTERPOSE to the path of the wrappers folder?

**Comment by jrfonseca** [comment]
No, you need to checkout the `dyld-interpose` branch, build, then run `apitrace trace ...` as usual.  It will setup `DYLD_INTERPOSE` correctly for you.

**Comment by lwjglapitrace** [comment]
So that gives me:

```
info: found /Users/lwjglapitrace/apitrace/build/wrappers/cgltrace.dylib
DYLD_INSERT_LIBRARIES=/Users/lwjglapitrace/apitrace/build/wrappers/cgltrace.dylib
java -Djava.library.path=libs/lwjgl-2.9.1/native/macosx/ -jar build/libs/app.jar
```
But still, no trace file is created :/

**Comment by jrfonseca** [comment]
That's odd.  No idea what could be happening.  Are you sure that app is really loading OpenGL?

I tried to repro this with lwjgl-2.9.3 but it just crashes with macOS 10.12, and I don't have 10.11 anymore.

**Comment by lwjglapitrace** [comment]
Yep, using apitrace on the same app on Linux works as expected :/

It's no big deal really, I can use my Linux box to debug the app, I was just curious if I was doing something wrong on OS X

**Comment by jrfonseca** [comment]
I got it working with LWJGL3 without any issues, using apitrace master branch:

    $ ~/projects/apitrace/apitrace trace -v ./runme-3.sh
    info: found /Users/jfonseca/projects/apitrace/wrappers/OpenGL.framework/OpenGL
    DYLD_FRAMEWORK_PATH=/Users/jfonseca/projects/apitrace/wrappers
    ./runme-3.sh
    + javac -cp lwjgl:lwjgl/res:lwjgl/jar/lwjgl.jar HelloWorld.java
    apitrace: loaded into /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/javac
    apitrace: unloaded from /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/javac
    + java -cp .:lwjgl/res:lwjgl/jar/lwjgl.jar -Djava.library.path=lwjgl/native -XstartOnFirstThread HelloWorld
    apitrace: loaded into /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/java
    Hello LWJGL 3.0.0 build 90!
    apitrace: tracing to /Users/jfonseca/projects/apitrace/issues/491/java.4.trace
    apitrace: attempting to read configuration file: /Users/jfonseca/Library/Preferences/apitrace/gltrace.conf

**Comment by leforg** [comment]
FWIW I also had this issue on OSX, specifically with Java

I was able to trace Blender runs, and apitrace successfully created the .trace files for me, but trying to create the trace files for a Java application failed

When I launched apitrace with the debug flag, I noticed some warnings around System Integrity Protection

So a Google for that brought me to here: http://www.macworld.com/article/2986118/security/how-to-modify-system-integrity-protection-in-el-capitan.html

And I followed the steps to disable SIP using csrutil in recovery mode, and since rebooting, .trace files are now created for Java applications

I have no idea why I needed to do this for Java in particular, and I'm not really clear on the security implications of disabling SIP, but I have .trace files now!
